### PR TITLE
if source and target owners are different, check quota on move

### PR DESCRIPTION
**Pull request by karakayasemi**
## Description
If source and target owners are different, check free space in dav quota plugin

## Related Issue
- Fixes https://github.com/owncloud/enterprise/issues/4427

## Motivation and Context
Fixing bugs

## How Has This Been Tested?
Manually tested below scenarios
- Create two users `user1` and `user2`
- Set quota for `user1` to unlimited, `user2` to 0
- Share a folder from `user1` to `user2`
- Upload a file from `user1` to shared folder, it should be successful
- `user1` moves the uploaded file to home folder, it should be successful
- Upload a file from `user1` to shared folder again, it should be successful
- When `user2` moves file to home folder, it should not be succesful and `user2` should show `Not enough free space` notification

Tried same scenario for chunked files also.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
I added issue #38592 to get some end-to-end test scenarios added. Those can be done in a later PR, if this PR is merged quickly.

File: apps/dav/lib/Connector/Sabre/QuotaPlugin.php [hunk 1]
```diff
@@ -80,27 +80,35 @@ public function initialize(\Sabre\DAV\Server $server) {
...
+			$targetNode = $this->server->tree->getNodeForPath($dirname);
 		}
 
+		// Check quota for only FutureFile or
+		// target Node owner different than source Node owner
+		if (!$sourceNode instanceof FutureFile) {
+			if (!$sourceNode instanceof Node || !$targetNode instanceof Node) {
+				return;
+			}
+			if ($sourceNode->getOwner()->getUID() === $targetNode->getOwner()->getUID()) {
```
Comment (jvillafanez) [inline apps/dav/lib/Connector/Sabre/QuotaPlugin.php:107]: Although unlikely, it might be possible that 2 different storages have the same owner.

Comment (VicDeo) [reply apps/dav/lib/Connector/Sabre/QuotaPlugin.php:107]: true. This case needs to be validated. e.g. move from local to external storage

Comment (karakayasemi) [reply apps/dav/lib/Connector/Sabre/QuotaPlugin.php:107]: We are handling these situations in storage wrappers, but I will try to cover in here also.

**Comment by phil-davis** [comment]
Note: I made some review comments about the tests in #38599 - when those are sorted out, the commit from that PR can be cherry-picked into here...

File: apps/dav/lib/Connector/Sabre/QuotaPlugin.php (hunk 1 above)
Comment (karakayasemi) [reply apps/dav/lib/Connector/Sabre/QuotaPlugin.php:109]: The described problem is also valid for external storages. Copying from external storage to different storage when encryption active and target storage leads to broken copy on target storage if target has not enough quota. 

I have covered this scenario in QuotaPlugin as you suggested.

Comment (jvillafanez) [reply apps/dav/lib/Connector/Sabre/QuotaPlugin.php:109]: I'd prefer to just check if the storage ids are different, as done lines below. I don't think we need to check the owners.

Comment (karakayasemi) [reply apps/dav/lib/Connector/Sabre/QuotaPlugin.php]: > 
> 
> I'd prefer to just check if the storage ids are different, as done lines below. I don't think we need to check the owners.

I agree. I have changed the code as you suggested.

**Review by phil-davis — COMMENTED** [review]
Tests look good now. dev review needed.

File: apps/dav/lib/Connector/Sabre/QuotaPlugin.php [hunk 2]
```diff
@@ -23,6 +23,7 @@
  *
  */
 namespace OCA\DAV\Connector\Sabre;
+use OC\Files\Storage\Home;
```
Comment (jvillafanez) [inline apps/dav/lib/Connector/Sabre/QuotaPlugin.php:26]: not used

**Review by jvillafanez — APPROVED** [review]
just a tiny detail

**Pull request merged**